### PR TITLE
[SPARK-23103][core] Ensure correct sort order for negative values in LevelDB.

### PR DESCRIPTION
**Pull request by vanzin**
The code was sorting "0" as "less than" negative values, which is a little
wrong. Fix is simple, most of the changes are the added test and related
cleanup.

### CODE REVIEW

File: common/kvstore/src/test/java/org/apache/spark/util/kvstore/DBIteratorSuite.java [hunk 1]
```diff
@@ -112,7 +114,8 @@ public void setup() throws Exception {
       t.key = "key" + i;
       t.id = "id" + i;
       t.name = "name" + RND.nextInt(MAX_ENTRIES);
-      t.num = RND.nextInt(MAX_ENTRIES);
+      // Force one item to have an integer value of zero to test the fix for SPARK-23103.
+      t.num = (i != 0) ? (int) RND.nextLong() : 0;
```
Comment (squito) [inline common/kvstore/src/test/java/org/apache/spark/util/kvstore/DBIteratorSuite.java]: why the chang from `RND.nextInt(MAX_ENTRIES)`?  this seems fine, just seemed like you were more likely to stress collision on this index before.

**Review by squito — APPROVED** [review]
lgtm

File: common/kvstore/src/test/java/org/apache/spark/util/kvstore/DBIteratorSuite.java (hunk 1 above)
Comment (vanzin) [reply common/kvstore/src/test/java/org/apache/spark/util/kvstore/DBIteratorSuite.java]: `nextInt` doesn't return negative values. I can add a mod operator here, but the code already explicitly generates some clashes a few lines below.

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
@vanzin The unit test fix is not related to the sorting issue right? Also it's totally fine to include it in this PR since it's minor.

**Comment by vanzin** [comment]
It is actually. Before, negative values (empty metrics all have value "-1") would show up when starting iteration from 0, and now they don't.

```
        store.view(classOf[TaskDataWrapper])
          .parent(stageKey)
          .index(TaskIndexNames.EXEC_RUN_TIME)
          .first(0L)
          .closeableIterator()
```

**Comment by jiangxb1987** [comment]
Ah, thanks for the explanation!

**Comment by squito** [comment]
even though we don't *know* of this causing a bug in 2.3, I still think we should merge it in there just because there may be some case we aren't thinking of, and this is a relatively small, safe fix.

so, I'm merging to master & 2.3

**Pull request closed without merging**